pca_cca.go

```diff
@@ -13,51 +13,96 @@ import (
-// an n×d matrix a where each row is an observation and each column is a
-// variable.
+// PC is the result of a principal components analysis. The results of a principal
+// components analysis held by a PC are only valid after a call to PrincipalComponents
+// has returned true.
+type PC struct {
+	n, d    int
+	weights []float64
+	svd     *mat64.SVD
+	ok      bool
+}
+
+// PrincipalComponents performs a weighted principal components analysis on the
+// matrix of the input data which is represented as an n×d matrix a where each
+// row is an observation and each column is a variable.
 //
 // PrincipalComponents centers the variables but does not scale the variance.
 //
 // The slice weights is used to weight the observations. If weights is nil, each
 // weight is considered to have a value of one, otherwise the length of weights
 // must match the number of observations or PrincipalComponents will panic.
 //
-// On successful completion, the principal component vectors are returned in the
-// columns of vecs as a d×min(n,d) matrix, and the column variances of the
-// principal component scores, b * vecs where b is a with centered columns, are
-// returned in vars as a min(n,d)-long slice in descending sort order.
-//
-// On failure, vecs and vars are returned nil and ok is returned false.
-func PrincipalComponents(a mat64.Matrix, weights []float64) (vecs *mat64.Dense, vars []float64, ok bool) {
-	n, d := a.Dims()
-	if weights != nil && len(weights) != n {
+// PrincipleCompoents returns true when the analysis is successful.
+func (c *PC) PrincipalComponents(a mat64.Matrix, weights []float64) (ok bool) {
+	c.n, c.d = a.Dims()
+	if weights != nil && len(weights) != c.n {
 		panic("stat: len(weights) != observations")
 	}
 
-	svd, ok := svdFactorizeCentered(a, weights)
-	if !ok {
-		return nil, nil, false
+	c.svd, c.ok = svdFactorizeCentered(c.svd, a, weights)
+	if c.ok {
+		c.weights = append(c.weights[:0], weights...)
 	}
+	return c.ok
+}
 
-	vecs = &mat64.Dense{}
-	vecs.VFromSVD(svd)
-	if n < d {
+// Vectors returns the component direction vectors of a principal components
+// analysis. The vectors are returned in the columns of a d×min(n,d) matrix.
+// If dst is not nil, it will be resized and used to store the direction vectors.
+// Vectors will panic if the receiver has not successfully performed a principal
+// components analysis.
+func (c *PC) Vectors(dst *mat64.Dense) *mat64.Dense {
+	if !c.ok {
+		panic("stat: use of unsuccessful principal components analysis")
+	}
+	if dst == nil {
+		dst = &mat64.Dense{}
+	} else {
+		// We are less strict here because SVD may retain more data
+		// than we may want, which we then discard below. Resizing
+		// correctly between repeated calls may be onerous.
+		dst.Reset()
+	}
+	dst.VFromSVD(c.svd)
+	if c.n < c.d {
 		// Don't retain columns that are not valid direction vectors.
-		vecs.Clone(vecs.Slice(0, d, 0, n))
+		dst = dst.Slice(0, c.d, 0, c.n).(*mat64.Dense)
```

Comment:
That is outstanding. Thank you. Yes it does have to change now - I can now be strict on sizing.

I blame myself for this weirdness.